design/DFX-Interface.md

```diff
@@ -99,22 +92,28 @@ This file informs Motoko about the interface of that canister. It could be the o
 
 Open problem: how to resolve mutual canister imports.
 
-Compiling IDL Files to JS
--------------------------
-
-In order to compile an IDL file, `dfx` invokes `didc` with
-
-    didc --js some/path/input.did -o another/path/output.js
-
-This _reads_ `some/path/input.did` and any `.did` file referenced by
-`some/path/input.did`.
-
-No constraints are imposed where these imported files reside (this may be refined to prevent relative imports from looking outside the project and the declared packages)
-
-This _writes_ to `another/path/output.js`, but has no other effect. It does
-not create `another/path/`.
-
-Compiler warnings and errors are reported to `stderr`. Nothing writes to `stdout`.
+Exporting Canister Metadata
+---------------------------
+
+The compiler generates various metadata about the canister via command line flags.
+The compiled Wasm module also includes these metadata in the custom sections.
+The compiler flag `--public-metadata <name>` controls if the custom section is publicly accessible.
+If `<name>` is in the `public-metadata` flag, the custom section name will be `icp:public <name>`.
+Otherwise, it will be `icp:private <name>`.
+
+* Candid interface.
+  + Compiler flag `--idl` generates the Candid interface for the canister. The main service
+    is always a service constructor, which contains the initialization arguments for installing the canister.
+  + Custom section `candid:service` stores the interface for the running (initialized) canister, which removes
+    the initializatio arguments.
+  + Custom section `candid:args` stores the initialization arguments. The argument types can refer to
+    types defined in the `candid:service` custom section.
+* Stable variable.
+  + Compiler flag `--print-stable-types` generates the signatures for stable variables.
```

Comment:
hmm, how permanent do we want to support importing actor class directly, instead of importing a Wasm? 

Seems to be a problem with all the flags we introduced here: `--public-metadata`, `--idl`, `--stable-types`. Do we want them to output all actor classes, or just the main one.

Reply:
> hmm, how permanent do we want to support importing actor class directly, instead of importing a Wasm?

Funnily enough, my original proposal for classes was just that, allowing you to import any wasm, whether produced by Motoko or not, but that idea got nixed for the more complicated idea of allowing recursive invocations of the compiler instead... One advantage of importing Motoko is that we can just use the Motoko types avoiding any ambiguity in the candid import (is that a principal, blob or vec nat8 (and word or nat in the past)?)

To be honest, I'm lukewarm on the whole public/private icp section distinction anyway: the IC doesn't enforce the accessibility on calls. Privacy is just making it a little harder for people to discover the interfaces, but once guessed or discovered, they can use and abuse them... I'm also finding it makes parsing custom section a little more awkward.... but I can live with that.

Reply:
For now, I think its simplest if we just treat the --public-metadata option as applying to all classes in the program...(and perhaps the others too, if that makes sense).

Reply:
Make sense for `--public-metadata`. 

For `--idl` and `--stable-types`, maybe better to just output the main service, and rely on `--print-deps` to generate other classes. If we apply these to all classes, moc will generate multiple did files at once. How does `dfx` know which files to look? The support for actor class in dfx is always vague though.